File: lib/core/version/app_version_service.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:package_info_plus/package_info_plus.dart';

import '../database/local_storage.dart';
import '../network/api_client.dart';
import 'app_version_models.dart';

class AppVersionService {
  AppVersionService(this._apiClient);

  final ApiClient _apiClient;

  static const _cacheBox = 'app_runtime';
  static const _cacheKey = 'app_version_policy_cache';
  static const _cachedAtKey = 'app_version_policy_cached_at';
  static const _defaultTtl = Duration(hours: 1);

  Future<AppVersionCheckResult> check() async {
    final info = await PackageInfo.fromPlatform();
    final installedVersion = info.version;

    if (kIsWeb) {
      return AppVersionCheckResult(
        decision: AppVersionDecision.upToDate,
        policy: const AppVersionPolicy(
          minVersion: '0.0.0',
          recommendedVersion: '0.0.0',
          storeUrlAndroid: '',
          storeUrlIos: '',
          messageForceUpdate: '',
          messageSoftUpdate: '',
          status: 'up_to_date',
          ttlSeconds: 3600,
        ),
        installedVersion: installedVersion,
      );
    }

    final platform = switch (defaultTargetPlatform) {
      TargetPlatform.android => 'android',
      TargetPlatform.iOS => 'ios',
      _ => 'android',
    };

    final cachedPolicy = await _loadCachedPolicy();
    final policy =
        cachedPolicy ?? await _fetchAndCachePolicy(platform, installedVersion);

    final decision = _decisionFromPolicy(policy, installedVersion);

    return AppVersionCheckResult(
      decision: decision,
      policy: policy,
      installedVersion: installedVersion,
    );
  }

  Future<AppVersionPolicy> _fetchAndCachePolicy(
    String platform,
    String appVersion,
  ) async {
    final response = await _apiClient.get<Map<String, dynamic>>(
      '/app/version',
      queryParameters: {'platform': platform, 'app_version': appVersion},
    );

    final envelope = response.data ?? <String, dynamic>{};
    final payload = (envelope['data'] is Map<String, dynamic>)
        ? envelope['data'] as Map<String, dynamic>
        : envelope;

    final policy = AppVersionPolicy.fromJson(payload);

    await LocalStorage.put<String>(
      _cacheBox,
      _cacheKey,
      jsonEncode(policy.toJson()),
    );
    await LocalStorage.put<int>(
      _cacheBox,
      _cachedAtKey,
      DateTime.now().millisecondsSinceEpoch,
    );

    return policy;
  }

  Future<AppVersionPolicy?> _loadCachedPolicy() async {
    final raw = await LocalStorage.get<String>(_cacheBox, _cacheKey);
    final cachedAt = await LocalStorage.get<int>(_cacheBox, _cachedAtKey);
    if (raw == null || cachedAt == null) {
      return null;
    }

    final ttl = DateTime.now().difference(
      DateTime.fromMillisecondsSinceEpoch(cachedAt),
    );
    if (ttl > _defaultTtl) {
      return null;
    }

    try {
      final decoded = jsonDecode(raw) as Map<String, dynamic>;
      return AppVersionPolicy.fromJson(decoded);
    } catch (_) {
      return null;
    }
  }

  AppVersionDecision _decisionFromPolicy(
    AppVersionPolicy policy,
    String installed,
  ) {
    if (_compareVersions(installed, policy.minVersion) < 0) {
      return AppVersionDecision.forceUpdate;
    }

    if (_compareVersions(installed, policy.recommendedVersion) < 0) {
      return AppVersionDecision.softUpdate;
    }

    return AppVersionDecision.upToDate;
  }

  int _compareVersions(String left, String right) {
    final l = _normalizeParts(left);
    final r = _normalizeParts(right);
    final max = l.length > r.length ? l.length : r.length;

    while (l.length < max) {
      l.add(0);
    }
    while (r.length < max) {
      r.add(0);
    }

    for (var i = 0; i < max; i++) {
      if (l[i] < r[i]) return -1;
      if (l[i] > r[i]) return 1;
    }

    return 0;
  }

  List<int> _normalizeParts(String version) {
    final clean = version.trim().replaceFirst(
      RegExp('^v', caseSensitive: false),
      '',
    );
    return clean.split('.').map((p) => int.tryParse(p) ?? 0).toList();
  }
}

File: backend/src/modules/app_version/app-version.service.ts
import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { AppVersionPolicy } from './entities/app-version-policy.entity';

type Platform = 'android' | 'ios';
type VersionStatus = 'force_update' | 'soft_update' | 'up_to_date';

@Injectable()
export class AppVersionService {
  constructor(
    @InjectRepository(AppVersionPolicy)
    private readonly repo: Repository<AppVersionPolicy>,
  ) {}

  async getPolicyForClient(platformRaw: string, appVersion?: string) {
    const platform = this.normalizePlatform(platformRaw);
    const policy = await this.repo.findOne({
      where: { platform, is_active: true },
      order: { updated_at: 'DESC' },
    });

    if (!policy) {
      throw new NotFoundException(`No active app version policy for ${platform}`);
    }

    const status = this.resolveStatus(appVersion, policy);

    return {
      platform,
      min_version: policy.min_version,
      recommended_version: policy.recommended_version,
      store_url_android: policy.store_url_android,
      store_url_ios: policy.store_url_ios,
      message_force_update: policy.message_force_update,
      message_soft_update: policy.message_soft_update,
      status,
      ttl_seconds: 3600,
      server_time: new Date().toISOString(),
    };
  }

  private normalizePlatform(value: string): Platform {
    const platform = (value ?? '').trim().toLowerCase();
    if (platform !== 'android' && platform !== 'ios') {
      throw new BadRequestException('platform must be android or ios');
    }
    return platform;
  }

  private resolveStatus(
    installedVersion: string | undefined,
    policy: AppVersionPolicy,
  ): VersionStatus {
    if (!installedVersion || !installedVersion.trim()) {
      return 'up_to_date';
    }

    const installed = this.normalizeVersion(installedVersion);
    const minVersion = this.normalizeVersion(policy.min_version);
    const recommendedVersion = this.normalizeVersion(policy.recommended_version);

    if (this.compareVersions(installed, minVersion) < 0) {
      return 'force_update';
    }

    if (this.compareVersions(installed, recommendedVersion) < 0) {
      return 'soft_update';
    }

    return 'up_to_date';
  }

  private normalizeVersion(raw: string): string {
    const clean = raw.trim().replace(/^v/i, '');
    const valid = /^\d+(\.\d+){0,3}$/.test(clean);
    if (!valid) {
      throw new BadRequestException(`Invalid version format: ${raw}`);
    }
    return clean;
  }

  private compareVersions(a: string, b: string): number {
    const aParts = a.split('.').map((part) => Number.parseInt(part, 10));
    const bParts = b.split('.').map((part) => Number.parseInt(part, 10));
    const maxLen = Math.max(aParts.length, bParts.length);

    for (let i = 0; i < maxLen; i += 1) {
      const left = aParts[i] ?? 0;
      const right = bParts[i] ?? 0;
      if (left < right) return -1;
      if (left > right) return 1;
    }

    return 0;
  }
}
